model_signing/serialization/dfs.py

```diff
@@ -0,0 +1,303 @@
+# Copyright 2024 The Sigstore Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Model serializers that build a single hash out of a DFS traversal."""
+
+import base64
+import concurrent.futures
+import pathlib
+from typing import Callable, Iterable, TypeAlias
+from typing_extensions import override
+
+from model_signing.hashing import file
+from model_signing.hashing import hashing
+from model_signing.manifest import manifest
+from model_signing.serialization import serialization
+
+
+def _check_file_or_directory(path: pathlib.Path) -> bool:
+    """Checks that the given path is either a file or a directory.
+
+    There is no support for sockets, pipes, or any other operating system
+    concept abstracted as a file.
+
+    Furthermore, this would return False if the path is a broken symlink, if it
+    doesn't exists or if there are permission errors.
+    """
+    if not (path.is_file() or path.is_dir()):
+        raise ValueError(
+            f"Cannot use '{path}' as file or directory. It could be a"
+            " special file, it could be missing, or there might be a"
+            " permission issue."
+        )
+
+
+def _build_header(
+    *,
+    entry_name: str,
+    entry_type: str,
+    start: int | None = None,
+    end: int | None = None,
+) -> bytes:
+    """Builds a header to encode a path with given name and type.
+
+    Args:
+        entry_name: The name of the entry to build the header for.
+        entry_type: The type of the entry (file or directory).
+    """
+    encoded_type = entry_type.encode("utf-8")
+    # Prevent confusion if name has a "." inside by encoding to base64.
+    encoded_name = base64.b64encode(entry_name.encode("utf-8"))
+
+    if start is not None and end is not None:
+        # Note: make sure to end with a ".".
+        encoded_range = f"{start}-{end}.".encode("utf-8")
+    else:
+        # Note: no "." at end here, it will be added by `join` on return.
+        encoded_range = b""
+
+    return b".".join([encoded_type, encoded_name, encoded_range])
```

Comment:
this will output `type.name.xxx-yyy`. I think we need an additional `.` at the end of the string, otherwise it's malleable `type.name.xxx-yyy + content == type.name.xxx-y + yycontent`. I think we're doing plain concatenation of header and conntent. Lmk if that's not the case

Reply:
I'm adding a dot at the end in line 65 (`{start}-{end}.`)

Reply:
might be more readable to add to the `return` line... but does not work with the empty encoded_range :/